extensions/amp-date-picker/0.1/amp-date-picker.js

```diff
@@ -0,0 +1,703 @@
+   */
+  renderTemplate_(template, opt_data, opt_fallback) {
+    if (template) {
+      const data = opt_data || /** @type {!JsonObject} */ ({});
+      return this.templates_.renderTemplate(template, data)
+          .then(rendered => {
+            rendered.setAttribute('i-amphtml-rendered', '');
+            const renderedEvent = createCustomEvent(
+                this.win_,
+                AmpEvents.DOM_UPDATE,
+                /* detail */ null,
+                {bubbles: true});
+
+            template.dispatchEvent(renderedEvent);
+            return rendered./*REVIEW*/outerHTML;
+          });
+    } else {
+      return Promise.resolve(opt_fallback || '');
+    }
+  }
+
+  /**
+   * Create the data needed to render a day template
+   * @param {!moment} date
+   * @return {!JsonObject}
+   */
+  getDayTemplateData_(date) {
+    const templateData = FORMAT_STRINGS.reduce((acc, key) => {
+      if (key === 'X' || key === 'x') {
+        const defaultLocaleDate = date.clone().locale(DEFAULT_LOCALE);
+        acc[key] = defaultLocaleDate.format(key);
+      } else {
+        acc[key] = date.format(key);
+      }
+      return acc;
+    }, {});
+    templateData.isHighlighted = this.highlighted_.contains(date);
+    templateData.isBlocked = this.blocked_.contains(date);
+
+    return templateData;
+  }
+
+  /**
+   * Render asynchronous HTML into a React component.
+   * @param {!Promise<string>} templatePromise
+   * @return {React.Component}
+   */
+  renderPromiseIntoReact_(templatePromise) {
+    if (!this.templateThen_) {
+      this.templateThen_ = html => this.react_.createElement('div', {
+        // This should be safe because this HTML is rendered through
+        // amp-mustache and is sanitized.
+        dangerouslySetInnerHTML: {__html: html},
+      });
+    }
+
+    return this.react_.createElement(createDeferred(this.react_), {
+      promise: templatePromise,
+      then: this.templateThen_,
+    });
+  }
+
+  /**
+   * Render the configured date picker component.
+   * @param {!Object=} opt_additionalProps
+   */
+  render(opt_additionalProps) {
+    const Picker = this.picker_;
+    const props = Object.assign({}, this.props_, opt_additionalProps);
+
+    if (this.hasDayTemplate_()) {
+      props.renderDay = this.renderDay;
+    }
+    if (this.hasInfoTemplate_()) {
+      props.renderCalendarInfo = this.renderInfo;
+    }
+
+    this.reactRender_(
+        this.react_.createElement(Picker, Object.assign({}, props, {
+          isRTL: this.isRTL_,
+          installActionHandler: this.installActionHandler_,
+          onDateChange: this.onDateChange,
+          onDatesChange: this.onDatesChange,
+          blocked: this.blocked_,
+          highlighted: this.highlighted_,
+          firstDayOfWeek: this.firstDayOfWeek_,
+          daySize: this.daySize_,
+        })),
+        this.container_);
+  }
+}
+
+
+/**
+ * Bootstraps the date picker elements
+ */
+export class AmpDatePickerService {
```

Comment:
I investigated making the date-picker an element with `layout="container"` like we discussed offline, and it looks like it works without any problems. So that resolves the main concern I had about making the date-picker a first-class AMP element